File: js-src/src/registries/Action.ts
import { runInAction } from "mobx";
import { getLogger } from "../logging";

const log = getLogger('Action library');

// Since mobx does not provide a way to register a global action middleware,
// we use a global action call stack to track the action nesting.
let _actionCallStack: Array<ActionState> = [];

// Middleware related
type ReportActionStateHook = (action: ActionState) => void;
let _rootActionStartedHooks: Array<ReportActionStateHook> = [];
let _rootActionCompletedHooks: Array<ReportActionStateHook> = [];

// Register the root action hooks
export function registerRootActionStartedHook(hook: ReportActionStateHook) {
    log.info('registerRootActionStartedHook called')
    _rootActionStartedHooks.push(hook);
}

export function registerRootActionCompletedHook(hook: ReportActionStateHook) {
    log.info('registerRootActionCompletedHook called')
    _rootActionCompletedHooks.push(hook);
}

function processMethod(descriptor: PropertyDescriptor, name: string, issuer: string): PropertyDescriptor {
    const originalMethod = descriptor.value; // Save the original method
    let funcName = `[${issuer}]${name}`;
    let func = function (this: any, ...args: any[]) {
        // Create a new action state
        let actionState = new ActionState(funcName);
        actionState.setStarted();

        if (_actionCallStack.length === 0) {
            // Call the root action hooks
            _rootActionStartedHooks.forEach(hook => hook(actionState));
        }
        _actionCallStack.push(actionState);

        // Call the original method
        let result: any;
        let exception: any = null;
        runInAction(() => {
            try{
                result = originalMethod.apply(this, args);
            } catch (e){
                log.error(`Error in action ${funcName}`, e);
                exception = e;
            }
        });
        if (exception) {
            // If an exception was thrown, set the action state to failed
            actionState.runState = ActionRunStates.FAILED;
        }


        // Complete the action state
        actionState = _actionCallStack.pop()!;
        actionState.setCompleted();

        if (_actionCallStack.length === 0) {
            // Call the root action hooks
            _rootActionCompletedHooks.forEach(hook => hook(actionState));
        }

        if (exception) {
            // If an exception was thrown, rethrow it
            throw exception;
        }

        return result;
    }
    // descriptor.value = mobxAction(funcName, func);
    descriptor.value = func;
    return descriptor;
}

// Overload interface for the action decorator
export interface IActionDecorator {
    // Overload for usage without arguments
    (target: Object, propertyKey: string | symbol, descriptor: PropertyDescriptor): void;

    // Overload for usage with options object
    (options: { name?: string, issuer?: string }): MethodDecorator;
}

export const action: IActionDecorator = function (...args: any[]): any {
    /**
     * A decorator to register a MobX action with additional metadata.
     *
     * This decorator can be applied with or without arguments. When used with arguments,
     * it allows specifying an issuer and an optional custom name for the action.
     * The custom name, if provided, replaces the method's original name in the MobX action context.
     * This is useful for middleware and services to distinguish user-triggered actions from
     * those triggered by services, aiding in functionalities like testing, undo/redo, etc.
     *
     * @param issuer - An optional issuer of the action, defaulting to 'user'.
     *                 Used to categorize the action's origin (e.g., user, system, service).
     * @param name - An optional custom name for the action. If provided, it replaces
     *               the method's original name in the action context.
     * @returns A decorator function that modifies the target method to be a MobX action.
     */

    // Check if it's being used as a decorator (without arguments)
    // console.log(`action decorator called with args: ${args}`)
    if (args.length === 3 && typeof args[0] === 'object') { // Normal method
        const [target, key, descriptor] = args as [Object, string | symbol, PropertyDescriptor];
        return processMethod(descriptor, key.toString(), 'user');
    } else if (args.length === 3 && typeof args[0] === 'function') { // Static method
        const [target, key, descriptor] = args as [Function, string | symbol, PropertyDescriptor];
        return processMethod(descriptor, key.toString(), 'user');
    } else if (args.length === 1 && typeof args[0] === 'object') {
        // There's an options dict supplied, which holds the named parameters
        const [options] = args as [{ name?: string, issuer?: string }];

        return function (target: Object, key: string | symbol, descriptor: PropertyDescriptor): PropertyDescriptor {
            const name = options.name || key.toString();
            const issuer = options.issuer || 'user';
            return processMethod(descriptor, name, issuer);
        }
    } else {
        console.log(args)
        throw new Error('Invalid usage of the action decorator');
    }
};


export enum ActionRunStates {
    NOT_STARTED = 'NOT_STARTED',
    STARTED = 'STARTED',
    COMPLETED = 'COMPLETED',
    FAILED = 'FAILED'
}


export class ActionState {
    name: string;
    runState: ActionRunStates;
    startTime: number;
    duration: number;

    constructor(name: string) {
        this.name = name;
        this.runState = ActionRunStates.NOT_STARTED;
        this.startTime = 0;
        this.duration = 0;
    }

    get issuer(): string {
        return this.name.split(']')[0].replace('[', '');
    }

    get started(): boolean {
        return this.runState === ActionRunStates.STARTED;
    }
    get completed(): boolean {
        return this.runState === ActionRunStates.COMPLETED;
    }

    setStarted() {
        this.runState = ActionRunStates.STARTED;
        this.startTime = Date.now();
    }

    setCompleted() {
        this.runState = ActionRunStates.COMPLETED;
        this.duration = Date.now() - this.startTime;
    }
    copy(): ActionState {
        let copy = new ActionState(this.name);
        copy.runState = this.runState;
        copy.startTime = this.startTime;
        copy.duration = this.duration;
        return copy;
    }
}

File: py-src/fusion/libs/action.py
import functools
import time
from enum import Enum
from typing import Callable

from fusion.logging import get_logger

log = get_logger(__name__)


class ActionRunStates(Enum):
    NOT_STARTED = "NOT_STARTED"
    STARTED = "STARTED"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"


class ActionState:
    """Carries the state of a single action invocation.
    Matches the TS ActionState class in fusion/registries/Action.ts.
    """

    def __init__(self, name: str):
        self.name = name
        self.run_state = ActionRunStates.NOT_STARTED
        self.start_time: float = 0
        self.duration: float = 0

    @property
    def issuer(self) -> str:
        """Parse issuer from name format '[issuer]method_name'."""
        return self.name.split("]")[0].replace("[", "")

    @property
    def started(self) -> bool:
        return self.run_state == ActionRunStates.STARTED

    @property
    def completed(self) -> bool:
        return self.run_state == ActionRunStates.COMPLETED

    def set_started(self):
        self.run_state = ActionRunStates.STARTED
        self.start_time = time.time()

    def set_completed(self):
        self.run_state = ActionRunStates.COMPLETED
        self.duration = time.time() - self.start_time

    def copy(self) -> "ActionState":
        c = ActionState(self.name)
        c.run_state = self.run_state
        c.start_time = self.start_time
        c.duration = self.duration
        return c

    def __repr__(self):
        return (
            f"<ActionState name={self.name} "
            f"run_state={self.run_state.value} "
            f"duration={self.duration * 1000:.2f}ms>"
        )


# Global action call stack to track nesting (matches TS _actionCallStack)
_action_call_stack: list[ActionState] = []

# Synchronous hook lists (matches TS _rootActionStartedHooks/_rootActionCompletedHooks)
_root_action_started_hooks: list[Callable] = []
_root_action_completed_hooks: list[Callable] = []


def register_root_action_started_hook(hook: Callable):
    log.info("register_root_action_started_hook called")
    _root_action_started_hooks.append(hook)


def register_root_action_completed_hook(hook: Callable):
    log.info("register_root_action_completed_hook called")
    _root_action_completed_hooks.append(hook)


def is_in_action() -> bool:
    return bool(_action_call_stack)


def check_in_action():
    """Guard for view state mutations — raises if not inside an @action."""
    if not is_in_action():
        raise Exception(
            "State changes can only happen in functions decorated with "
            "the @action decorator"
        )


def action(name: str, issuer: str = "user"):
    """Decorator that wraps a function as an action with call-stack tracking
    and synchronous root-action hooks.

    Matches the TS processMethod() flow in Action.ts.

    Args:
        name: The action name (e.g. 'window.close_tab')
        issuer: Origin of the action ('user' or 'service')
    """
    if not name or not isinstance(name, str):
        raise Exception(
            "Please add the action name as an argument to the decorator. "
            "E.g. @action('action_name')"
        )

    func_name = f"[{issuer}]{name}"

    def decorator(func):
        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            # Create action state and mark started
            action_state = ActionState(func_name)
            action_state.set_started()

            # If root action — fire started hooks synchronously
            if len(_action_call_stack) == 0:
                for hook in _root_action_started_hooks:
                    hook(action_state)

            _action_call_stack.append(action_state)

            # Call the original function
            exception = None
            result = None
            try:
                result = func(*args, **kwargs)
            except Exception as e:
                log.error(f"Error in action {func_name}: {e}")
                exception = e

            if exception:
                action_state.run_state = ActionRunStates.FAILED

            # Pop and complete
            _action_call_stack.pop()
            if action_state.run_state != ActionRunStates.FAILED:
                action_state.set_completed()
            else:
                action_state.duration = time.time() - action_state.start_time

            # If root action — fire completed hooks synchronously
            if len(_action_call_stack) == 0:
                for hook in _root_action_completed_hooks:
                    hook(action_state)

            if exception:
                raise exception

            return result

        return wrapper

    return decorator
